DRILL-7951: Add Response Metadata Fields to HTTP Storage Plugin

cgivre opened the pull request:
# [DRILL-7951](https://issues.apache.org/jira/browse/DRILL-7951): Add Response Metadata Fields to HTTP Storage Plugin

## Description

In many http calls, the response code can contain meaningful information, however these fields are not available unless the call results in an error.

This PR adds four implicit metadata fields to the HTTP storage plugin which are:
* `_response_code`: The HTTP response code. This field is returned as an INT.
* `_response_message`: The response message.
* `_response_protocol`: The response protocol.
* `_response_url`: The actual URL that is sent to the API.

As these are implicit fields, they will not appear in star queries.

## Documentation
See above.

## Testing
Added three additional unit tests.

luocooong commented:
@cgivre Thanks for the PR. Is it ready to review?

cgivre commented:
> @cgivre Thanks for the PR. Is it ready to review?

I'm not sure why these tests are failing as they don't have anything to do with the changes I made.  Let me rerun them a few times to see if it is a CI issue or an actual issue.

cgivre commented:
> > @cgivre Thanks for the PR. Is it ready to review?
> 
> I'm not sure why these tests are failing as they don't have anything to do with the changes I made. Let me rerun them a few times to see if it is a CI issue or an actual issue.

@luocooong 
Now this is ready for review. :-). 

@jfmario, Do you have any suggestions?

luocooong commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/ImplicitColumnUtils.java:
```diff
@@ -0,0 +1,141 @@
+     */
+    private int addImplicitColumnToSchema(String fieldName, MinorType type, RowSetLoader rowWriter) {
+      ColumnMetadata colSchema = MetadataUtils.newScalar(fieldName, type, DataMode.OPTIONAL);
+      colSchema.setBooleanProperty(ColumnMetadata.EXCLUDE_FROM_WILDCARD, true);
```
I am interested in the usage of `setBooleanProperty(ColumnMetadata.EXCLUDE_FROM_WILDCARD, true)`.

luocooong commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/ImplicitColumnUtils.java:
```diff
@@ -0,0 +1,141 @@
+    private final Map<String, ImplicitColumn> implicitColumns;
+    private final RowSetLoader rowWriter;
+
+    public ImplicitColumns (RowSetLoader rowWriter) {
```
Remove the white space after the `ImplicitColumns`.

cgivre replied:
Fixed

luocooong commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/ImplicitColumnUtils.java:
```diff
@@ -0,0 +1,141 @@
+      this.dataType = dataType;
+      this.fieldName = fieldName;
+      this.columnIndex = addImplicitColumnToSchema(this.fieldName, this.dataType, rowWriter);
+      this.writer = rowWriter.scalar(this.columnIndex);
```
Should we check that the index already exists?
```java
index = rowWriter.tupleSchema().index(fieldName);
if (index == -1) {
  index = addImplicitColumnToSchema(this.fieldName, this.dataType, rowWriter);
}
```

luocooong commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/loader/JsonLoaderImpl.java:
```diff
@@ -276,6 +285,10 @@ public boolean readBatch() {
     while (rowWriter.start()) {
       if (parser.next()) {
+        // Add implicit fields
+        if (implicitFields != null) {
```
I think the `implicitColumns` may be empty (Not null). So, I recommend that
```java
if(!implicitColumns.isEmpty()) {
  // do that
}
```

cgivre replied:
In the next line it calls the `writeImplicitColumns()` method which has a `foreach` loop.  If the implicit columns are empty (not null) it will skip the loop and return.

luocooong reviewed (commented):
@cgivre Thanks for the revision. There are minor things...
In addition, The value of the local variable `previousState` is not used (line 445, in XMLReader).

jfmario commented:
@cgivre I think a lot of this looks good. I noticed that when I query an API and receive a 404 response with no body, Drill throws an error instead of giving me those fields.

I think there may be cases where that is preferred, but my view is that a 404 with no body is an informative response and not an error, so I would have preferred to see the "fact of" the 400 response in the returning table. The same thing applies to other 400 errors.

This may be something that users want to configure. I think 500s are always errors, but if you add a configuration it may be worth letting users specify which status codes should be treated as errors.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpBatchReader.java:
```diff
@@ -24,26 +24,34 @@
 
 public class HttpBatchReader implements ManagedReader<SchemaNegotiator> {
+
+  private static final String[] STRING_METADATA_FIELDS = {"_response_message", "_response_protocol", "_response_url"};
```
I think the extra whitespace after the `{` violates the Drill style sheet.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpBatchReader.java:
```diff
@@ -97,6 +112,21 @@ public void addContext(UserException.Builder builder) {
+  }
+
+  protected void populateImplicitFieldMap(SimpleHttp http) {
+    implicitColumns.getColumn("_response_message").setValue(http.getResponseMessage());
```
Fixed

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpXMLBatchReader.java:
```diff
@@ -80,8 +81,14 @@ public void addContext(UserException.Builder builder) {
+
       RowSetLoader rootRowWriter = resultLoader.writer();
       xmlReader.open(rootRowWriter, errorContext);
+      xmlReader.setMetadata(implicitColumns);
```
Renamed for consistency.

cgivre commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/util/SimpleHttp.java:
```diff
@@ -171,6 +175,12 @@ public InputStream getInputStream() {
+      responseMessage = response.message();
+      responseCode = response.code();
+      responseProtocol = response.protocol().toString();
+      responseURL = response.request().url().toString();
```
I'm fairly certain that by definition these fields will be populated.  For instance, in order to get a response, you have to send data to a URL, hence that URL will be defined in the response.  I added a null check and the IDE complained and confirmed that `protocol` and the `url` field will never be null.

cgivre commented:
Hi @luocooong, @jfmario 
Thank you for the review comments.   @luocooong I think I addressed all your comments.

@jfmario 
I added a configuration option to the API Connection, `errorOn400` which allows the user to decide whether they want Drill to throw an error or not on 400 response codes.

luocooong reviewed (approved):
LGTM. +1
